alerts/proxy_drop_executable.py

```diff
@@ -0,0 +1,83 @@
+#!/usr/bin/env python
+
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+# Copyright (c) 2014 Mozilla Corporation
+
+
+from lib.alerttask import AlertTask
+from query_models import QueryStringMatch, SearchQuery, TermMatch, ExistsMatch
+
+
+class AlertProxyDropExecutable(AlertTask):
+    def main(self):
+
+        superquery = None
+        run = 0
+
+        self.parse_config('proxy_drop_executable.conf', [
+                          'destinationfilter', 'extensions'])
+
+        search_query = SearchQuery(minutes=5)
+
+        search_query.add_must([
+            TermMatch('category', 'squid'),
+            TermMatch('tags', 'squid'),
+            TermMatch('details.proxyaction', 'TCP_DENIED/-')
+        ])
+
+        search_query.add_must_not([
+            QueryStringMatch(
+                'details.destination: /{}/'.format(self.config.destinationfilter)),
+        ])
+
+        # TODO: remove HACK in onAggregation once this works as expected
+        # for extension in self.config.extensions.split(','):
+        #     if run == 0:
+        #         superquery = QueryStringMatch(
+        #             'details.destination: /{0}/'.format(extension))
+        #     else:
+        #         superquery |= QueryStringMatch(
+        #             'details.destination: /{0}/'.format(extension))
+        #     run += 1
+        #
+        # search_query.add_must(superquery)
+
+        self.filtersManual(search_query)
+
+        # Search aggregations on field 'hostname', keep X samples of
+        # events at most
+        self.searchEventsAggregated('details.sourceipaddress', samplesLimit=10)
+        # alert when >= X matching events in an aggregation
+        # I think it makes sense to alert every time here
+        self.walkAggregations(threshold=1)
+
+    # Set alert properties
+    def onAggregation(self, aggreg):
+        # aggreg['count']: number of items in the aggregation, ex: number of failed login attempts
+        # aggreg['value']: value of the aggregation field, ex: toto@example.com
+        # aggreg['events']: list of events in the aggregation
+        category = 'squid'
+        tags = ['squid', 'proxy']
+        severity = 'WARNING'
+
+        dropped_url_destinations = []
+
+        # START HACK - The conditionals here are to account for an inability to endwith
```

Comment:
This hack is implemented to show we can make the unit-tests pass without fixing the query search.  The concern however is that by not fixing this we can run into a situation where a given source IP has multiple deny events within an aggregated search and the ones we want are discarded (or not sampled) and then we further weed those events in the onAggregation such that a exe target failure would basically not fire in some circumstances.  Not to mention, it's horribly unperformant to do this in python and not in Lucene/ES (at least in concept, I'm not sure how much more overhead this presents).